src/Rule.php

```diff
@@ -41,22 +47,25 @@ abstract class Rule
      */
     abstract protected function validateValue($value, DataSetInterface $dataSet = null): Result;
 
-    public function setTranslator(TranslatorInterface $translator): self
+    public function withTranslator(TranslatorInterface $translator): self
     {
-        $this->translator = $translator;
-        return $this;
+        $new = clone $this;
```

Comment:
Do you really need to create a new copy? I don't think there is a use-case. If your goal is cascading, you should:
```php
public function translator(TranslatorInterface $translator): self 
{
  $this->translator = $translator;
  return $this;
}

$validator->translator(...)->translatorDomain(...)->translationLocale(...)->...
```

Reply:
@kamarton cloning is cheap. I don't think there's anything bad in making object immutable if it's not meant to have state.